laurent22/joplin
# Mobile: Support importing from JEX files

# Summary

Related forum post: https://discourse.joplinapp.org/t/import-jex-file-on-android/31126/8

This pull request adds support for importing from .JEX files on mobile. 

# Notes

This pull request is mostly UI changes (`InteropService.instance().import` already works for `.jex` files on mobile). Specifically, it:
1. Renames the "Export" section in settings to "Import and Export"
2. Creates a new `TaskButton.tsx` file that handles displaying progress and warnings for both export and import tasks.
3. Adds a new import button that uses `TaskButton`.

Note that it now also possible to do multiple exports without closing and re-opening the export settings tab.

Currently, imported notebooks are added to the top level (not added within a notebook). It may make sense to change this, for example, by creating a new notebook for each import.

# Screenshots

Note: The `foo` warning was added as a demonstration of the warning UI:
![Screenshot: Import complete, shows one warning directly below the "Import" button, "Completed with warnings: \ foo"](https://github.com/laurent22/joplin/assets/46334387/e994c13c-41b3-4341-bfad-60b9f4de3601) ![screenshot: Import and export screen: Import running, shows progress bar, button labeled "Importing"](https://github.com/laurent22/joplin/assets/46334387/e5df5570-de74-465f-8417-ba0815a9b48a)
![screenshot: Main settings screen in dark mode, the renamed "Import and Export" tab is highlighted.](https://github.com/laurent22/joplin/assets/46334387/8833280e-974f-474b-9a9b-c574133130b5)


# Testing

1. Ensure that at least one image and emoji or other non-ASCII character is present in a note.
1. Export all  notes as JEX from the "Import and Export" settings screen.
2. Import from the just-exported JEX file.
3. Verify that notebooks and notes have been duplicated.
4. Verify that the image and emoji from step 1 are present in the duplicated (imported) note.

This has been tested successfully on an Android 12 emulator.
